cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -304,7 +299,10 @@ public void processMessage(MetadataMessage message) {
       }
       DatasetUsage usage = message.getPayload(GSON, DatasetUsage.class);
       ProgramId programId = (ProgramId) message.getEntityId();
-      usageDataset.register(programId, usage.getDatasetId());
+      TransactionRunners.run(transactionRunner, context -> {
```

Comment:
I believe all the table operation in this class do not require to be wrapped in the TransactionRunner, the `processMessages` is used in a way it is already wrapped in a transaction. You can construct the table using the `StructuredTableContext` in the parameter of `processMessages`. The same apply to all the usage of the `transactionRunner`.

Reply:
Hmm good point, done